### PR TITLE
feat: Currency Manager

### PR DESCRIPTION
**Pull request by benjlevesque**
# Breaking change!

Introduce a flexible way to manage the list of currencies, so that each application can add their own currencies.
Provide a unique currency manager, that also supports adding custom meta information to a currency.

## The list of currencies is no longer globally shared:
- it is passed as a property of the `RequestNetwork` client
- it is optional and defaults to the current list

## `Currency` no longer exists: 
- it is replaced by a `CurrencyManager` that contains a list of currencies and provide ways to access one currency. 
- It can be extended to add custom metadata
```ts
import { CurrencyManager } from "@requestnetwork/currencies";

type Meta = { usdRate: number  };
const currencyManager = new CurrencyManager<Meta>([{
  address: "0x...",
  symbol: "DAI",
  decimals: 18,
  meta: { usdRate: 1.01 }
  type: "ERC20"
}]);
const currency = currencyManager.from("DAI");
console.log(currency) // { symbol: "DAI", decimals: 18, [...] meta: { usdRate: 1.01 } }
```
- it no longer throws when currency is not found. It returns `undefined`

## Representation of a currency

A `CurrencyInput` represents the minimal required information. It is the input of the `CurrencyManager`
A `CurrencyDefinition` represents the complete information: `CurrencyInput` + calculated fields (unique `id` and `hash`)

### CODE REVIEW

File: packages/request-client.js/test/index.test.ts [hunk 1]
```diff
@@ -1873,4 +1873,107 @@ describe('index', () => {
...
+            currency: '_TEST',
+          },
+          paymentNetwork,
+          signer: payeeIdentity,
+        });
+
+        expect(request.getData().currency).toBe('_TEST');
+      });
+
+      it('does not contain a default token', async () => {
```
Comment (yomarion) [inline packages/request-client.js/test/index.test.ts:1968]: I don't understand this test

**Review by yomarion — APPROVED** [review]
Great job !

File: packages/request-client.js/src/api/request.ts [hunk 2]
```diff
@@ -78,6 +83,7 @@ export default class Request {
   constructor(
     requestId: RequestLogicTypes.RequestId,
     requestLogic: RequestLogicTypes.IRequestLogic,
+    currencyManager: ICurrencyManager,
```
[the remark this answers is not in the document]
Comment (benjlevesque) [reply packages/request-client.js/src/api/request.ts:86]: we never implement directly this object, it's always implemented from the RequestNetwork client. I don't see a good reason to handle defaults at several different places?

File: packages/payment-detection/src/payment-network-factory.ts [hunk 3]
```diff
@@ -59,11 +60,13 @@ export default class PaymentNetworkFactory {
     currency,
     paymentNetworkCreationParameters,
     bitcoinDetectionProvider,
+    currencyManager,
   }: {
     advancedLogic: AdvancedLogicTypes.IAdvancedLogic;
     currency: RequestLogicTypes.ICurrency;
     paymentNetworkCreationParameters: PaymentTypes.IPaymentNetworkCreateParameters;
     bitcoinDetectionProvider?: PaymentTypes.IBitcoinDetectionProvider;
+    currencyManager: ICurrencyManager;
```
[the remark this answers is not in the document]
Comment (benjlevesque) [reply packages/payment-detection/src/payment-network-factory.ts:73]: same answer as before, why handle defaults in different places? isn't it better to reduce the # of times we set the defaults?

File: packages/payment-detection/src/erc20/fee-proxy-contract.ts [hunk 4]
```diff
@@ -39,13 +41,21 @@ export default class PaymentNetworkERC20FeeProxyContract<
...
   /**
    * @param extension The advanced logic payment network extensions
    */
-  public constructor({ advancedLogic }: { advancedLogic: AdvancedLogicTypes.IAdvancedLogic }) {
+  public constructor({
+    advancedLogic,
+    currencyManager,
+  }: {
+    advancedLogic: AdvancedLogicTypes.IAdvancedLogic;
+    currencyManager: ICurrencyManager;
```
[the remark this answers is not in the document]
Comment (benjlevesque) [reply packages/payment-detection/src/erc20/fee-proxy-contract.ts:55]: same as before

File: packages/request-client.js/test/index.test.ts (hunk 1 above)
Comment (benjlevesque) [reply packages/request-client.js/test/index.test.ts]: better now?

**Review by kevindavee — APPROVED** [review]

**Review by AdamDowson — APPROVED** [review]
Awesome

**Pull request merged**